debugger.md

```diff
@@ -0,0 +1,160 @@
+# Debugger proposal
+
+## Background
+
+One of the greatest advantages of being a log based workflow engine is that Temporal workflows are replayable.
+
+A developer can download workflow history and run it through the SDK replayer to materialize the workflow program's
+state.
+
+Workflows, being "just code", can be debugged with tools the developer is familiar with.
+
+## Features
+
+### Overview
+
+Typically, when looking at workflow history, it's clear which workflow task a failure occurs in but it's not immediately
+obvious which line of code that correlates to.
+
+This proposal discusses how we can make the debugging experience better for developers.
+
+### Supported languages
+
+Initially we will support Core based languages (TypeScript and Python) but Go and Java could also be supported with
+additional effort.
+
+The main advantage of going for Core based languages is that those language SDKs run Rust Core in a separate thread
+pool, allowing interaction with the runtime while a workflow is paused by a debugger.
+
+Targeting Core based SDKs will allow us to reuse a lot of the code for Ruby, .NET and possibly other SDKs if we decide
+to rewrite them to be backed by Core.
+
+### Supported IDEs
+
+Initially we will support Microsoft's Visual Studio Code only while keeping in mind other IDEs.
+
+The rationale being that at least for JavaScript VSCode is by far the most popular IDE and we'd get the most value
+investing in providing good DX for that environment.
+
+### Components
+
+- VSCode plugin
+  - Main panel / Workflow history panel
+  - Debug adapter
+- Core replay debug server
+- Lang middleware
+
+#### Main panel
+
+Has input to enter workflow ID and optional run ID with submit button to start a debug session.
+Alternatively can have a file (path) input to use a history JSON or binary file (as exported by web and tctl or
+downloaded with a client).
+
+The credentials for connecting to a server should be stored in the system keychain (can use
+[keytar](https://www.npmjs.com/package/keytar) to manage that).
+
+#### Workflow history panel
+
+Once a debug session has started, the main panel switches to history view.
+In this view, the history is displayed in a format with this rough guideline:
+
+```
+[ ] Workflow Task
+> Event in 1
+> Event in 2
+< Command out 1
+< Command out 2
+```
+
+`[ ]` - is a breakpoint toggle
+
+The current workflow task will be highlighed when being processed.
+
+#### Debug adapter
+
+Use to step over SDK code and customized per language
+
+#### Core replay debug server
+
+A webserver opened by Core, listens on port allocated by the VSCode plugin.
+Has a single endpoint for setting breakpoints on workflow task started events (by event ID).
+
+In order to indicate to Lang that an activation has a breakpoint associated with it, we'll add a new field to
+[`WorkflowActivation`][workflow-activation]
```

Comment:
I am a bit confused about this point. Why is lang involved? Isn't the current workflow task info delivered on a side channel? And isn't whether something has hit a breakpoint or not between core and vscode? I don't think lang has anything do to here. I do understand if the vscode side has to step out until they get to user code.

Reply:
Communication between core and vscode is async.
After thinking about this some more, we can do all of this with minimal lang interference and no changes to core.
We could intercept activations and notify vscode when SDK starts to process each workflow task, vscode could use that information to set a breakpoint if required.

Reply:
I think you still need some small change to core. It doesn't need to be specific to this, though.

Right now lang doesn't know which workflow task it's on. More than one activation can exist for one workflow task. This problem can be solved by attaching some kind of "workflow task ID" to each activation, though, and then it's easy to know when you've moved from one WFT to the next.

Unless I'm just not following what this was supposed to be for in the first place.

Reply:
I was thinking we can use this: https://github.com/temporalio/sdk-core/blob/15ea5d3ee580e2a226eefdcf0d8444b2ae30a997/protos/local/temporal/sdk/core/workflow_activation/workflow_activation.proto#L30

Reply:
That is probably stable, but, a bit of semantic abuse. It's not hard to add something more appropriate to core.